### PR TITLE
PDFExporter v2.0

### PR DESCRIPTION
**Pull request by david-livadaru-3pillar**

### CODE REVIEW

**Comment by adriansomesan** [comment]
:+1:

File: Example/PDFExporterDemo/PDFExporterDemo/Controllers/SettingsScreen/SettingsViewController.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 @property (weak, nonatomic) id<SettingsViewControllerDelegate> delegate;
 
-@property (nonatomic) UIEdgeInsets contentPaperInsets;
+@property (nonatomic) UIEdgeInsets contentPaperInset;
```
Comment (vladrusu-3pillar) [inline Example/PDFExporterDemo/PDFExporterDemo/Controllers/SettingsScreen/SettingsViewController.h]: could be a typo here, as the type is UIEdgeInsets and the name of the property is Inset

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITextView+PDFExporterDrawing.m [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+//  PDFExporter
+//
+//  Copyright © 2016 3Pillar Global. All rights reserved.
+//
+
+#import "UITextView+PDFExporterDrawing.h"
+
+@implementation UITextView (PDFExporterDrawing)
+
+- (BOOL)canDrawSubview:(UIView *)subview intersection:(CGRect)intersection {
```
Comment (vladrusu-3pillar) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UITextView+PDFExporterDrawing.m]: Would be nice if cases like this one were documented, it's hard to understand what's happening

**Comment by vladrusu-3pillar** [comment]
:+1:

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        return;
+    }
+    UIScrollViewPersistenceState *scrollState = (UIScrollViewPersistenceState *)state;
+    self.contentOffset = scrollState.contentOffset;
+}
+
+#pragma mark - PDFExporterPageInformation
+
+- (CGRect)subviewRect:(UIView *)subview layoutPageRect:(CGRect)rect {
+    if ([subview isKindOfClass:[UITableViewCell class]]) {  // UITableView's cells are wrapped inside a UIScrollView
```
Comment (sean7512) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m]: this is a big assumption and Apple can change this "expected" behavior at any time.

You should check if superview exists and is a scrollview instead?

FYI: I am a US 3Pillar employee

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    if (self.layer.borderWidth != 0) {
+        [self drawBorderWithPath:path];
+    }
+}
+
+- (void)drawSubviewsWithPath:(UIBezierPath *)path withinPageRect:(CGRect)rect {
+    CGPoint offset = CGPointZero;
+    for (CGFloat yCoordinate = CGRectGetMinY(rect);
+         yCoordinate < CGRectGetMaxY(rect);
+         yCoordinate += CGRectGetHeight(self.bounds)) {
```
Comment (sean7512) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: No need to call CGRectGetMaxY and CGRectGetHeight for each iteration

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        offset.y = yCoordinate;
+        self.contentOffset = offset;
+        CGRect frame = self.frame;
+        [self layoutIfNeeded];
+        self.frame = frame;
```
Comment (sean7512) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: Do you expect `self.frame` to change after layout?  If it does change, it ends up being overridden back to the original frame anyways.  Why is this necessary?

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+         yCoordinate < CGRectGetMaxY(rect);
+         yCoordinate += CGRectGetHeight(self.bounds)) {
+        
+        // prepare view for rect section
+        contentOffset.y = yCoordinate;
+        self.contentOffset = contentOffset;
+        CGRect frame = self.frame;
+        [self layoutIfNeeded];
+        self.frame = frame;
+        [self layoutHeadersAndFooters];
```
Comment (sean7512) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: this chunk of code is duplicated and should be pulled out

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UIView+PDFExporterStatePersistance.m [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+@interface UIView (PDFExporterStatePersistancePrivate)
+
+@property (nonatomic) NSMutableArray<UIViewPersistenceState *> *persistenceStates;
+
+@end
+
+@implementation UIView (PDFExporterStatePersistance)
+
+- (NSMutableArray<UIViewPersistenceState *> *)persistenceStates {
```
Comment (sean7512) [inline Framework/PDFExporter/PDFExporter/Categories/UIKit/UIView+PDFExporterStatePersistance.m]: :+1: For using generics...it would be nice to use them everywhere as well.  Makes Swift use a lot easier

**Comment by sean7512** [comment]
Would be nice to see the API audited for Nullability for easier use in Swift

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m (hunk 3 above)
Comment (gabrieldusa) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m]: This is a good idea. I also think we should add this check.

Comment (david-livadaru-3pillar) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m]: subview.superview is self in this case. The check is redundant.

File: Example/PDFExporterDemo/PDFExporterDemo/Controllers/SettingsScreen/SettingsViewController.h (hunk 1 above)
Comment (david-livadaru-3pillar) [reply Example/PDFExporterDemo/PDFExporterDemo/Controllers/SettingsScreen/SettingsViewController.h]: As per Apple naming convention, we use the singular form (e.g.: contentInset from `UIScrollView`, separatorInset from `UITableViewCell`).

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m (hunk 4 above)
Comment (david-livadaru-3pillar) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: What is happening here is that we're looping from minimum Y to maximum Y with using the height of bounds as step.

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITextView+PDFExporterDrawing.m (hunk 2 above)
Comment (david-livadaru-3pillar) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UITextView+PDFExporterDrawing.m]: We have written documentation for source code.

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m (hunk 5 above)
Comment (david-livadaru-3pillar) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: It is not needed anymore. It has been removed.

File: Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m (hunk 6 above)
Comment (david-livadaru-3pillar) [reply Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m]: We have removed the duplicated code.

**Review by ciprian-antal-3PG — APPROVED** [review]
